src/Runner/Example.hs

```diff
@@ -13,52 +13,118 @@ data Result = Equal | NotEqual [String]
   deriving (Eq, Show)
 
 mkResult :: ExpectedResult -> [String] -> Result
-mkResult expected actual
-  | expected `matches` actual = Equal
-  | otherwise = NotEqual (formatNotEqual expected actual)
+mkResult expected_ actual_ =
+  case expected `matches` actual of
+  Nothing         -> Equal
+  Just (col, row) -> NotEqual (formatNotEqual expected actual col row)
   where
-    chunksMatch :: [LineChunk] -> String -> Bool
-    chunksMatch [] "" = True
-    chunksMatch [LineChunk xs] ys = stripEnd xs == stripEnd ys
+    -- use show to escape special characters in output lines if any output line
+    -- contains any unsafe character
+    escapeOutput
+      | any (not . isSafe) $ concat (expectedAsString ++ actual_) = init . tail . show . stripEnd
+      | otherwise = id
+
+    actual :: [String]
+    actual = escapeOutput <$> actual_
+
+    expected :: ExpectedResult
+    expected = transformExcpectedLine escapeOutput <$> expected_
+
+    expectedAsString :: [String]
+    expectedAsString = map (\x -> case x of
+        ExpectedLine str -> concatMap lineChunkToString str
+        WildCardLine -> "..." ) expected_
+
+    isSafe :: Char -> Bool
+    isSafe c = c == ' ' || (isPrint c && (not . isSpace) c)
+
+    chunksMatch :: [LineChunk] -> String -> Maybe (String, String)
+    chunksMatch [] "" = Nothing
+    chunksMatch [LineChunk xs] ys =
+      if stripEnd xs == stripEnd ys
+      then Nothing
+      else matchingPrefix xs ys
     chunksMatch (LineChunk x : xs) ys =
-        x `isPrefixOf` ys && xs `chunksMatch` drop (length x) ys
+      if x `isPrefixOf` ys
+      then (\(m, p) -> (x ++ m, x ++ p)) <$> (xs `chunksMatch` drop (length x) ys)
+      else matchingPrefix x ys
     chunksMatch zs@(WildCardChunk : xs) (_:ys) =
-        xs `chunksMatch` ys || zs `chunksMatch` ys
-    chunksMatch _ _ = False
+      case xs `chunksMatch` ys of
+      Nothing  -> Nothing
+      Just (m, col) -> case zs `chunksMatch` ys of
+                       Nothing   -> Nothing
+                       Just (m2, col2) ->
+                         let (mm, coll) = if length m2 > length m then (m2, col2) else (m, col) in
+                         Just $ (mm, '.':coll)
+    chunksMatch [WildCardChunk] [] = Nothing
+    chunksMatch (WildCardChunk:_) [] = Just ("", "")
+    chunksMatch [] (_:_) = Just ("", "")
 
-    matches :: ExpectedResult -> [String] -> Bool
-    matches (ExpectedLine x : xs) (y : ys) = x `chunksMatch` y && xs `matches` ys
-    matches (WildCardLine : xs) ys | xs `matches` ys = True
-    matches zs@(WildCardLine : _) (_ : ys) = zs `matches` ys
-    matches [] [] = True
-    matches [] _  = False
-    matches _  [] = False
+    matchingPrefix xs ys =
+      let a = fst <$> takeWhile (\(x, y) -> x == y) (xs `zip` ys) in
+      Just (a, a)
+
+    matches :: ExpectedResult -> [String] -> Maybe (String, Int)
```

Comment:
Instead of a tuple, how about defining a record type? If not, please add a comment describing what each field is.